### PR TITLE
Parallel

### PR DESCRIPTION
**Pull request by diogoaihara**
- Executions in parallel for the same destination type
- Reads only once per source name

### CODE REVIEW

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py [hunk 1]
```diff
@@ -21,50 +21,52 @@
...
-from models.execution import TransactionalType
+from models.execution import TransactionalType, ExecutionsGroupedBySource
 
 _BIGQUERY_PAGE_SIZE = 20000
 
 _LOGGER_NAME = 'megalista.data_sources.BigQuery'
 
 class BigQueryDataSource(BaseDataSource):
-    def __init__(self, transactional_type: TransactionalType, bq_ops_dataset: str, bq_location: str, source_type: SourceType, source_name: str, destination_type: DestinationType, destination_name: str):
-        self._transactional_type = transactional_type
```
[the remark this answers is not in the document]
Comment (diogoaihara) [reply megalista_dataflow/data_sources/big_query/big_query_data_source.py:36]: This part has been transferred into BaseDataSource constructor

File: megalista_dataflow/data_sources/base_data_source.py [hunk 2]
```diff
@@ -11,11 +11,23 @@
...
 # See the License for the specific language governing permissions and
 # limitations under the License.
-from models.execution import Execution
-from typing import Any, Iterable
+from site import execsitecustomize
+from models.execution import Execution, ExecutionsGroupedBySource, TransactionalType, DataRowsGroupedBySource
+from typing import Any, Iterable, List
+from apache_beam.typehints.decorators import with_output_types
 
 class BaseDataSource:
```
[the remark this answers is not in the document]
Comment (diogoaihara) [reply megalista_dataflow/data_sources/base_data_source.py:19]: That's a good idea, but I think it's better to do it in another PR.

File: megalista_dataflow/data_sources/big_query/big_query_data_source.py [hunk 3]
```diff
@@ -118,17 +132,21 @@ def _ensure_control_table_exists(self, client: Client, uploaded_table_name: str)
         client.query(query).result()
 
     def write_transactional_info(self, rows, execution: Execution):
-        table_name = self._get_table_name(execution.source.source_metadata, True)
+        if len(rows) == 0:
+            logging.getLogger(_LOGGER_NAME).info(
+                "No rows to insert. Skipping...")
```
Comment (Mr-Lopes) [inline megalista_dataflow/data_sources/big_query/big_query_data_source.py:137]: We could include self._source_name within this skip message

File: megalista_dataflow/data_sources/file/file_data_source.py [hunk 4]
```diff
@@ -169,30 +169,50 @@ def _update_dtypes(self, destination_type: DestinationType, col_names: list) ->
...
-
+    def _get_data_frame_from_file(self, file: io.BytesIO, is_uploaded: bool) -> pd.DataFrame:
+        destination_type = self._destination_type
+        if is_uploaded:
+            destination_type = DestinationType.UPLOADED_GCLID_TIME if self._transactional_type == TransactionalType.GCLID_TIME else DestinationType.UPLOADED_UUID
+        cols = pd.read_csv(file, dtype='string', nrows=0).columns
+        if DataSchemas.validate_data_columns(cols, destination_type):
+            cols = DataSchemas.get_cols_names(cols, destination_type)
+            file.seek(0)
+            logging.getLogger(_LOGGER_NAME).info(cols)
```
Comment (Mr-Lopes) [inline megalista_dataflow/data_sources/file/file_data_source.py:209]: Do we need to log the columns here?

**Review by Mr-Lopes — APPROVED** [review]
Approved with a few comments :)
This is a great improvement for Megalista!

**Pull request merged**